File: src/main/java/net/arvandor/numinoustreasury/stamina/NuminousStaminaService.java
package net.arvandor.numinoustreasury.stamina;

import com.rpkit.characters.bukkit.character.RPKCharacter;
import com.rpkit.characters.bukkit.character.RPKCharacterService;
import com.rpkit.core.service.Service;
import com.rpkit.core.service.Services;
import com.rpkit.players.bukkit.profile.minecraft.RPKMinecraftProfile;
import com.rpkit.players.bukkit.profile.minecraft.RPKMinecraftProfileService;
import net.arvandor.numinoustreasury.NuminousTreasury;
import org.bukkit.entity.Player;
import org.jooq.DSLContext;

import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;

public final class NuminousStaminaService implements Service {

    private final NuminousTreasury plugin;
    private final NuminousCharacterStaminaRepository staminaRepository;
    private final Map<Integer, Integer> characterStamina;

    public NuminousStaminaService(NuminousTreasury plugin, NuminousCharacterStaminaRepository staminaRepository) {
        this.plugin = plugin;
        this.staminaRepository = staminaRepository;
        this.characterStamina = new ConcurrentHashMap<>();
    }

    @Override
    public NuminousTreasury getPlugin() {
        return plugin;
    }

    public int getStamina(RPKCharacter character) {
        return characterStamina.get(character.getId().getValue());
    }

    public int getStamina(Player player) {
        RPKCharacter character = getRpkCharacter(player);
        if (character == null) return 0;
        return getStamina(character);
    }

    public void getAndUpdateStamina(Player player, StaminaUpdateFunction function, StaminaUpdateCallback callback) {
        RPKCharacter character = getRpkCharacter(player);
        if (character == null) return;
        getAndUpdateStamina(character, function, callback);
    }

    public void getAndUpdateStamina(RPKCharacter character, StaminaUpdateFunction function, StaminaUpdateCallback callback) {
        plugin.getServer().getScheduler().runTaskAsynchronously(plugin, () -> {
            staminaRepository.getAndUpdate(character.getId(), function, (oldStamina, newStamina) -> {
                characterStamina.put(character.getId().getValue(), newStamina);
                plugin.getServer().getScheduler().runTask(plugin, () -> {
                    callback.invoke(oldStamina, newStamina);
                });
            });
        });
    }

    public void setStamina(Player player, int stamina, Runnable callback) {
        RPKCharacter character = getRpkCharacter(player);
        if (character == null) return;
        setStamina(character, stamina, callback);
    }

    public void setStamina(RPKCharacter character, int stamina, Runnable callback) {
        plugin.getServer().getScheduler().runTaskAsynchronously(plugin, () -> {
            staminaRepository.setStamina(character.getId(), stamina);
            characterStamina.put(character.getId().getValue(), stamina);
            plugin.getServer().getScheduler().runTask(plugin, callback);
        });
    }

    public void setStamina(DSLContext dsl, Player player, int stamina) {
        RPKCharacter character = getRpkCharacter(player);
        if (character == null) return;
        setStamina(dsl, character, stamina);
    }

    public void setStamina(DSLContext dsl, RPKCharacter character, int stamina) {
        staminaRepository.setStamina(dsl, character.getId(), stamina);
        characterStamina.put(character.getId().getValue(), stamina);
    }

    public void restoreStamina(Runnable callback) {
        plugin.getServer().getScheduler().runTaskAsynchronously(plugin, () -> {
           staminaRepository.restoreStamina();
           characterStamina.clear();
           plugin.getServer().getScheduler().runTask(plugin, () -> {
               List<RPKCharacter> onlineCharacters = plugin.getServer().getOnlinePlayers().stream()
                       .map(this::getRpkCharacter)
                       .filter(Objects::nonNull)
                       .toList();
               plugin.getServer().getScheduler().runTaskAsynchronously(plugin, () -> {
                   onlineCharacters.forEach(this::load);
                   plugin.getServer().getScheduler().runTask(plugin, callback);
               });
           });
        });
    }

    private RPKCharacter getRpkCharacter(Player player) {
        RPKMinecraftProfileService minecraftProfileService = Services.INSTANCE.get(RPKMinecraftProfileService.class);
        if (minecraftProfileService == null) return null;
        RPKMinecraftProfile minecraftProfile = minecraftProfileService.getPreloadedMinecraftProfile(player);
        if (minecraftProfile == null) return null;
        RPKCharacterService characterService = Services.INSTANCE.get(RPKCharacterService.class);
        if (characterService == null) return null;
        return characterService.getPreloadedActiveCharacter(minecraftProfile);
    }

    public void load(RPKCharacter character) {
        characterStamina.put(character.getId().getValue(), staminaRepository.getStamina(character.getId()));
    }

    public void unload(RPKCharacter character) {
        characterStamina.remove(character.getId().getValue());
    }

}

File: src/main/kotlin/net/arvandor/numinoustreasury/stamina/NuminousStaminaService.kt
package net.arvandor.numinoustreasury.stamina

import com.rpkit.characters.bukkit.character.RPKCharacter
import com.rpkit.characters.bukkit.character.RPKCharacterService
import com.rpkit.core.service.Service
import com.rpkit.core.service.Services
import com.rpkit.players.bukkit.profile.minecraft.RPKMinecraftProfileService
import net.arvandor.numinoustreasury.NuminousTreasury
import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player
import org.jooq.DSLContext
import java.util.concurrent.ConcurrentHashMap

class NuminousStaminaService(
    private val plugin: NuminousTreasury,
    private val staminaRepository: NuminousCharacterStaminaRepository,
) : Service {
    private val characterStamina: MutableMap<Int, Int> = ConcurrentHashMap()
    val maxStamina: Int = plugin.config.getInt("stamina.max")

    override fun getPlugin(): NuminousTreasury {
        return plugin
    }

    fun getStamina(character: RPKCharacter): Int {
        return characterStamina[character.id.value] ?: 0
    }

    fun getStamina(player: Player): Int {
        val character = getRpkCharacter(player) ?: return 0
        return getStamina(character)
    }

    fun getAndUpdateStamina(
        player: OfflinePlayer,
        function: StaminaUpdateFunction,
        callback: StaminaUpdateCallback,
    ) {
        val character = getRpkCharacter(player) ?: return
        getAndUpdateStamina(character, function, callback)
    }

    fun getAndUpdateStamina(
        character: RPKCharacter,
        function: StaminaUpdateFunction,
        callback: StaminaUpdateCallback,
    ) {
        plugin.server.scheduler.runTaskAsynchronously(
            plugin,
            Runnable {
                staminaRepository.getAndUpdate(character.id, function) { oldStamina: Int, newStamina: Int ->
                    characterStamina[character.id.value] = newStamina
                    plugin.server.scheduler.runTask(
                        plugin,
                        Runnable {
                            callback.invoke(oldStamina, newStamina)
                        },
                    )
                }
            },
        )
    }

    fun setStamina(
        player: OfflinePlayer,
        stamina: Int,
        callback: Runnable?,
    ) {
        val character = getRpkCharacter(player) ?: return
        setStamina(character, stamina, callback)
    }

    fun setStamina(
        character: RPKCharacter,
        stamina: Int,
        callback: Runnable? = null,
    ) {
        plugin.server.scheduler.runTaskAsynchronously(
            plugin,
            Runnable {
                staminaRepository.setStamina(character.id, stamina)
                characterStamina[character.id.value] = stamina
                if (callback != null) {
                    plugin.server.scheduler.runTask(plugin, callback)
                }
            },
        )
    }

    fun setStamina(
        dsl: DSLContext,
        player: OfflinePlayer,
        stamina: Int,
    ) {
        val character = getRpkCharacter(player) ?: return
        setStamina(dsl, character, stamina)
    }

    fun setStamina(
        dsl: DSLContext,
        character: RPKCharacter,
        stamina: Int,
    ) {
        staminaRepository.setStamina(dsl, character.id, stamina)
        characterStamina[character.id.value] = stamina
    }

    fun restoreStamina(callback: Runnable? = null) {
        plugin.server.scheduler.runTaskAsynchronously(
            plugin,
            Runnable {
                staminaRepository.restoreStamina()
                characterStamina.clear()
                plugin.server.scheduler.runTask(
                    plugin,
                    Runnable {
                        val onlineCharacters =
                            plugin.server.onlinePlayers.mapNotNull { player: OfflinePlayer ->
                                this.getRpkCharacter(
                                    player,
                                )
                            }
                        plugin.server.scheduler.runTaskAsynchronously(
                            plugin,
                            Runnable {
                                onlineCharacters.forEach { character -> this.load(character) }
                                if (callback != null) {
                                    plugin.server.scheduler.runTask(plugin, callback)
                                }
                            },
                        )
                    },
                )
            },
        )
    }

    private fun getRpkCharacter(player: OfflinePlayer): RPKCharacter? {
        val minecraftProfileService =
            Services.INSTANCE.get(
                RPKMinecraftProfileService::class.java,
            )
        if (minecraftProfileService == null) return null
        val minecraftProfile = minecraftProfileService.getPreloadedMinecraftProfile(player) ?: return null
        val characterService =
            Services.INSTANCE.get(
                RPKCharacterService::class.java,
            )
        if (characterService == null) return null
        return characterService.getPreloadedActiveCharacter(minecraftProfile)
    }

    fun load(character: RPKCharacter) {
        characterStamina[character.id.value] = staminaRepository.getStamina(character.id)
    }

    fun unload(character: RPKCharacter) {
        characterStamina.remove(character.id.value)
    }
}
